dashboard/app/models/contact_rollups_pardot_memory.rb

```diff
@@ -26,21 +26,73 @@
 class ContactRollupsPardotMemory < ApplicationRecord
   self.table_name = 'contact_rollups_pardot_memory'
 
-  # Retrieves new email-Pardot ID mappings from Pardot and saves them to the database.
-  # @param [Integer, nil] last_id retrieves only Pardot ID greater than this value
-  def self.add_and_update_pardot_ids(last_id = nil)
+  # Downloads and saves new email-Pardot ID mappings from Pardot.
+  # *Warning:* This method overwrites existing data.
+  #
+  # @param [Integer] last_id retrieves only Pardot ID greater than this value
+  # @param [Integer] limit the maximum number of Pardot prospects to download
+  def self.download_pardot_ids(last_id = nil, limit = nil)
     last_id ||= ContactRollupsPardotMemory.maximum(:pardot_id) || 0
+    fields = %w(id email)
 
-    PardotV2.retrieve_new_ids(last_id) do |mappings|
+    PardotV2.retrieve_prospects(last_id, fields, limit) do |prospects|
       current_time = Time.now.utc
-      batch = mappings.map {|item| item.merge(pardot_id_updated_at: current_time)}
+      batch = prospects.map do |item|
+        {
+          email: item['email'],
+          pardot_id: item['id'].to_i,
+          pardot_id_updated_at: current_time
+        }
+      end
 
       import! batch,
         validate: false,
         on_duplicate_key_update: [:pardot_id, :pardot_id_updated_at]
     end
   end
 
+  # Downloads and saves prospect data from Pardot.
+  # *Warning:* This method overwrites existing data.
```

Comment:
clarification q: does it wipe the existing data and replace it with whatever is downloaded here or do some sort of partial overwrite?

Reply:
It upserts (inserts + updates) downloaded data into contact_rollups_pardot_memory table. Only the update operation will overwrite existing data.